### PR TITLE
Use Persistent Key-Val DB for Shard Node Collation Storage

### PR DESCRIPTION
**Pull request by rauljordan**
Hi all,

This PR references #109 in order to use a persistent, key-value db as a shard backend that is stored to a local folder in the user's `DataDir` called `shardchaindata` as geth currently does with the `eth` and `les` clients.

## Requirements for Merge

- [x] Generalize the ShardBackend interface and ensure tests pass
- [x] Implement levelDB as ShardBackend
- [x] Ensure information is persisted on disk even between sessions, write tests for this

### CODE REVIEW

File: sharding/shard_test.go [hunk 1]
```diff
@@ -64,7 +65,7 @@ func TestShard_HeaderByHash(t *testing.T) {
 	header := NewCollationHeader(big.NewInt(1), &emptyHash, big.NewInt(1), &emptyAddr, []byte{})
 
 	// creates a mockDB that always returns nil values from .Get and errors in every other method.
-	mockDB := &mockShardDB{kv: make(map[common.Hash]*[]byte)}
+	mockDB := &mockShardDB{kv: make(map[common.Hash][]byte)}
 
 	// creates a well-functioning shardDB.
 	shardDB := database.NewShardKV()
```
Comment (terenc3t) [inline sharding/shard_test.go]: shouldn't this be `NewShardDB()`?

File: sharding/shard_test.go [hunk 2]
```diff
@@ -346,7 +347,7 @@ func TestShard_SetAvailability(t *testing.T) {
 	header := NewCollationHeader(big.NewInt(1), &chunkRoot, big.NewInt(1), nil, []byte{})
 
 	// creates a mockDB that always returns nil values from .Get and errors in every other method.
-	mockDB := &mockShardDB{kv: make(map[common.Hash]*[]byte)}
+	mockDB := &mockShardDB{kv: make(map[common.Hash][]byte)}
 
 	// creates a well-functioning shardDB.
 	shardDB := database.NewShardKV()
```
Comment (terenc3t) [inline sharding/shard_test.go]: shouldn't this be `NewShardDB()`?

File: sharding/database/database.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	Has(k []byte) (bool, error)
+	Put(k []byte, val []byte) error
+	Delete(k []byte) error
+}
+
+// NewShardDB initializes a shardDB that writes to local disk.
+func NewShardDB(dataDir string, name string) (ShardBackend, error) {
+	// Uses default cache and handles values.
+	// TODO: allow these arguments to be set based on cli context.
+	return ethdb.NewLDBDatabase(filepath.Join(dataDir, name), 16, 16)
```
[the remark this answers is not in the document]
Comment (rauljordan) [reply sharding/database/database.go]: Yeah, we can start an issue for that.

**Comment by prestonvanloon** [comment]
Travis is failing linter and cannot build tests

**Pull request merged**